### PR TITLE
User backend extendable account searching

### PR DESCRIPTION
**Pull request by tomneedham**
## Description
Continued from: https://github.com/owncloud/core/pull/27832

### Problem
 - In OC9 we allow LDAP users to specify custom search attributes which can be used for finding users to share
 - In OC10 we have the accounts table which is used for searching to reduce the hits to LDAP servers
 - Organisations use these search attributes to help users find other users based on terms they may know (like usernames, emails, staff codes, etc)

### Proposed / Implemented Solution
 - Added a new oc_account_terms table with `account_id` (foreign key to accounts.id) and `term` columns
 - Altered user manager code to find users using these terms
 - Allow user backends to manipulate these search terms as needed (e.g: LDAP can add these from custom attributes the user configures)

User_LDAP App:
 - PR in progress to implement these features and supply the search attributes to the core https://github.com/owncloud/user_ldap/pull/98

## Related Issue
https://github.com/owncloud/user_ldap/issues/96

## How Has This Been Tested?
 - Manual testing - needs more 💪  
 - Updated unit tests -> still WIP 🚧 

## Screenshots (if appropriate):
because they make your PR look more cool:
screen shot 2017-05-16 at 15 44 52

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/User/AccountTerm.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ */
+
+namespace OC\User;
+
+
+use OCP\AppFramework\Db\Entity;
+
+/**
+ * Class Account
```
Comment (PVince81) [inline lib/private/User/AccountTerm.php]: AccountTerm

**Review by PVince81 — COMMENTED** [review]
At first look the code looks fine

**Comment by butonic** [comment]
The idea of the account table was to only rely on eg ldap for the login to be able to iterate over all users etc. The code that used to lead to a lot of ldap queries because it would execute pages etc was the search method used when sharing users. But also fetching any other use metadata was triggering ldap queries. Unfortunately, the ldap user bachend allos specifying additional fields that should be taken into account when doing a user search. that might eg be the `city` or multi value attributes like `mailAccounts`. That is actually a nice feature. The initial approach does not fit as @tomneedham and I today by chance stumbled over an LDIF entry with 7 email addresses. We were told there were accounts with many more. We came up with the idea of an additional table to hold arbitrary search terms. Another approach would be to add an elasticsearch based framework for this kind of cacheable metadata ... but that is nothing we can add on short notice.

For now I think this is a good approach. We could add a foreign key but did not want to be the first ones to actually use the relevant code. We would have liked to have an ORMapper, but got around it with adding an interface and making the ldap app actively update the search attributes on first login.

**Comment by DeepDiver1975** [comment]
> Hitting the LDAP server when someone types in a share search string might be ok especially as it benefits from LDAP specific searching capabilities (and patterns?).

this is one major complaint I'm hearing from big installations: searching the ldap takes ages. From an UX pov this is not acceptable.

File: lib/private/User/Account.php [hunk 2]
```diff
@@ -86,4 +88,29 @@ public function getBackendInstance() {
...
+	 * @param string[] $terms
+	 */
+	public function setSearchTerms(array $terms) {
+		if(array_diff($terms, $this->terms)) {
+			$this->terms = $terms;
+			$this->_termsChanged = true;
+		}
+	}
+
+	public static function fromRow(array $row) {
```
Comment (DeepDiver1975) [inline lib/private/User/Account.php]: needed?

File: lib/private/User/Account.php [hunk 3]
```diff
@@ -65,6 +64,9 @@ class Account extends Entity {
 	protected $state;
 	protected $home;
 
+	protected $terms = [];
+	private $_termsChanged = false;
```
Comment (DeepDiver1975) [inline lib/private/User/Account.php]: _ ?? needed ???

File: lib/private/User/AccountTermMapper.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	public function __construct(IDBConnection $db) {
+		parent::__construct($db, 'account_terms', AccountTerm::class);
+	}
+	/**
+	 * Sets search terms for a given account id, will always be lowercased
+	 * @param $account_id
+	 * @param string[] $terms
+	 */
+	public function setTermsForAccount($account_id, array $terms) {
```
Comment (DeepDiver1975) [inline lib/private/User/AccountTermMapper.php]: use Mapper:insert ???

File: lib/private/User/AccountTermMapper.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			$qb->setParameter('term', strtolower($term));
+			$qb->execute();
+		}
+	}
+
+	/**
+	 * Removes all search terms for a given account id
+	 * @param $account_id
+	 */
+	public function deleteTermsForAccount($account_id) {
```
Comment (DeepDiver1975) [inline lib/private/User/AccountTermMapper.php]: use Mapper:delete ???

File: lib/private/User/Manager.php [hunk 6]
```diff
@@ -170,7 +176,7 @@ protected function getUserObject(Account $account, $cacheUser = true) {
 			return $this->cachedUsers[$account->getUserId()];
 		}
 
-		$user = new User($account, $this->accountMapper, $this, $this->config, null, \OC::$server->getEventDispatcher() );
+		$user = new User($account, $this->accountMapper, $this->accountTermMapper, $this, $this->config, null, \OC::$server->getEventDispatcher() );
```
Comment (DeepDiver1975) [inline lib/private/User/Manager.php]: is the terms mapper necessary inside class User? I'd hide the terms mapper within the account mapper

File: lib/public/IGroupManager.php [hunk 7]
```diff
@@ -124,6 +124,18 @@ public function getUserGroupIds($user, $scope = null);
 	public function displayNamesInGroup($gid, $search = '', $limit = -1, $offset = 0);
 
 	/**
+	 * search for users in a specific group
+	 *
+	 * @param string $gid
+	 * @param string $search
+	 * @param int $limit
+	 * @param int $offset
+	 * @return array an array of display names (value) and user ids (key)
```
Comment (DeepDiver1975) [inline lib/public/IGroupManager.php]: not the User objects?

File: tests/lib/Group/ManagerTest.php [hunk 8]
```diff
@@ -1137,4 +1137,339 @@ public function testGroupDisplayName() {
...
+					case 'user33': return true;
+					default:
+						return null;
+				}
+			}));
+
+		/**
+		 * @var \OC\User\Manager $userManager
+		 */
+		$userManager = $this->createMock('\OC\User\Manager');
```
Comment (DeepDiver1975) [inline tests/lib/Group/ManagerTest.php]: Manager:class

File: tests/lib/Group/ManagerTest.php [hunk 9]
```diff
@@ -1137,4 +1137,339 @@ public function testGroupDisplayName() {
...
+		$userBackend = $this->createMock('\OC_User_Backend');
```
Comment (DeepDiver1975) [inline tests/lib/Group/ManagerTest.php]: OC_User_Backend::class

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by tomneedham** [comment]
Rebased

File: lib/private/User/Account.php (hunk 3 above)
Comment (butonic) [inline lib/private/User/Account.php]: _ to indicate it is used like _updatedFields: tracking if the attribute was changed so the mapper can act accordingly

File: lib/private/User/AccountTermMapper.php (hunk 5 above)
Comment (butonic) [inline lib/private/User/AccountTermMapper.php]: we want to delete by account_id, not id

File: lib/private/User/AccountTermMapper.php (hunk 4 above)
Comment (butonic) [inline lib/private/User/AccountTermMapper.php]: adressed with https://github.com/owncloud/core/pull/27906/commits/dd377c94b645bc50d3e365ec500aafa72f507002#diff-1e60c46513705f729f8e728292eb6f3bR43

**Comment by PVince81** [comment]
This doesn't work with guest emails (haven't tried generic emails), steps:

1. Login as admin
1. Enable guest app
1. Create folder "test1"
1. Type "Vince" in share field for "test1", select "Share to guest"
1. Type in local email address, "vincent@host.tld"
1. Log out
1. Open mail, open link and set password
1. Login as admin again
1. Create another folder "test2"
1. Share "test2" with "vincent@host.tld"

Expected: autocomplete finds the guest user "Vince"
Actual: autocomplete doesn't find the user and only offers a "remote" user (due to the format).

This worked on the previous PR.

@tomneedham did you properly include searching by email ? I can see that the oc_accounts column "email" for that guest user contains the correct email address. (and it's unique)

**Comment by PVince81** [comment]
Fixed by replacing "inner join" with "left join": https://github.com/owncloud/core/pull/27906/commits/06d7c36c66d18a888f318c7f8dcd55f77749c54c

**Comment by tomneedham** [comment]
🤞 for jenkins ;)

**Comment by butonic** [comment]
investigating the index question in https://github.com/owncloud/core/pull/27832#discussion_r117225157 made me look up how ldap handles it. turns out we have another config option hidden in the code: https://github.com/owncloud/user_ldap/blob/5e2a75bb78f57b8a406467ad48ea68faa856a85d/lib/Access.php#L1368-L1368

It is disabled by default, so we get the same search behavior as with ldap.

we use individual indexes because every db but sqlite can combine them and they help with `getByEmail` and `getByUid` as well.

@tomneedham @DeepDiver1975 @PVince81 please review.

**Comment by PVince81** [comment]
that option was supposed to be an unsupported experimental feature, do we need it enabled now ?

**Comment by PVince81** [comment]
@jvillafanez

File: lib/private/User/AccountMapper.php [hunk 10]
```diff
@@ -205,7 +220,7 @@ public function callForAllUsers($callback, $search, $onlySeen) {
 		$qb->select(['*'])
 			->from($this->getTableName());
 
-		if ($search) {
+		if ($search) { // TODO use medial search config option here as well
```
Comment (PVince81) [inline lib/private/User/AccountMapper.php]: something for much later ?

**Comment by butonic** [comment]
no we dont need it enabled, but since it was there i assumed someone was using it.

File: lib/private/User/AccountMapper.php (hunk 10 above)
Comment (butonic) [inline lib/private/User/AccountMapper.php]: well if the medial search was only experimental we can get rid of all the TODOs as well as the config option stuff.

Comment (PVince81) [inline lib/private/User/AccountMapper.php]: hmm, yeah... if the search is running entirely within OC tables then the option is not useful any more.

It was added for people with special requirements but not default due to perf impact.

**Comment by PVince81** [comment]
please test extensively in the RC2, thanks

**Comment by mrow4a** [comment]
Hey guys, what was the motivation to add setter and getter of search terms to account https://github.com/owncloud/core/blob/master/lib/private/User/Account.php#L104-L104 ? Is that really needed? I cannot find nowhere in the code sth using it @tomneedham

**Comment by tomneedham** [comment]
Since our entity ORM stuff doesnt deal with parent/child relationships we had to half bake this in ourselves. When an account is first created it is used to store search terms: https://github.com/owncloud/core/blob/master/lib/private/User/Manager.php#L452 and also when a user is synchronised: https://github.com/owncloud/core/blob/master/lib/private/User/SyncService.php#L174 

(but this logic is under a TODO to combine it since it is duplicated)

**Comment by tomneedham** [comment]
group search terms is another idea.....

**Pull request merged**